probe/src/flash/flasher.rs

```diff
@@ -0,0 +1,583 @@
+        let r = f(&mut active)?;
+        active.uninit()?;
+        Ok(r)
+    }
+
+    pub fn flash_block(
+        mut self,
+        address: u32,
+        data: &[u8],
+        chip_erase: Option<bool>,
+        smart_flash: bool,
+        fast_verify: bool,
+    ) -> Result<(), FlasherError> {
+        if !self.region.range.contains_range(&(address..address + data.len() as u32)) {
+            return Err(FlasherError::AddressNotInRegion(address, self.region.clone()));
+        }
+
+        let mut fb = FlashBuilder::new(self.region.range.start);
+        fb.add_data(address, data).expect("Add Data failed");
+        fb.program(self, chip_erase, smart_flash, fast_verify, true).expect("Add Data failed");
+
+        Ok(())
+    }
+
+        // def flash_block(self, addr, data, smart_flash=True, chip_erase=None, progress_cb=None, fast_verify=False):
+        // """!
+        // @brief Flash a block of data.
+        // """
+        // assert self.region is not None
+        // assert self.region.contains_range(start=addr, length=len(data))
+        
+        // fb = FlashBuilder(self)
+        // fb.add_data(addr, data)
+        // info = fb.program(chip_erase, progress_cb, smart_flash, fast_verify)
+        // return info
+}
+
+pub struct ActiveFlasher<'a, O: Operation> {
+    session: &'a mut Session,
+    region: &'a FlashRegion,
+    double_buffering_supported: bool,
+    _operation: core::marker::PhantomData<O>,
+}
+
+impl<'a, O: Operation> ActiveFlasher<'a, O> {
+    pub fn region(&self) -> &FlashRegion {
+        &self.region
+    }
+
+    pub fn flash_algorithm(&self) -> &FlashAlgorithm {
+        &self.session.target.info.flash_algorithm
+    }
+
+    pub fn session_mut(&mut self) -> &mut Session {
+        &mut self.session
+    }
+
+    pub fn uninit<'b, 's: 'b>(&'s mut self) -> Result<Flasher<'b>, FlasherError> {
+        log::debug!("Running uninit routine.");
+        let algo = self.session.target.info.flash_algorithm;
+
+        if let Some(pc_uninit) = algo.pc_uninit {
+            let result = self.call_function_and_wait(
+                pc_uninit,
+                Some(O::operation()),
+                None,
+                None,
+                None,
+                false
+            )?;
+
+            if result != 0 {
+                return Err(FlasherError::Uninit(result));
+            }
+        }
+
+        Ok(Flasher {
+            session: self.session,
+            region: self.region,
+            double_buffering_supported: self.double_buffering_supported,
+        })
+    }
+
+    fn call_function_and_wait(&mut self, pc: u32, r0: Option<u32>, r1: Option<u32>, r2: Option<u32>, r3: Option<u32>, init: bool) -> Result<u32, FlasherError> {
+        self.call_function(pc, r0, r1, r2, r3, init)?;
+        Ok(self.wait_for_completion())
+    }
+
+    fn call_function(&mut self, pc: u32, r0: Option<u32>, r1: Option<u32>, r2: Option<u32>, r3: Option<u32>, init: bool) -> Result<(), FlasherError> {
```

Comment:
What exactly does the init parameter do?
And maybe the 3 Option arguments could be replaced with `&[u32]` or an Enum, like this the
code gets a bit unreadable.

Reply:
I'll do that in an issue :)